### PR TITLE
Added option for azure authentication

### PR DESCRIPTION
**Pull request by cblims**
Needed a way to authenticate to azure, so I added an option for that. Also removed the code that checks for email address and issuerUrl, as this information is already in the jwt token.

### CODE REVIEW

File: cmd/auth.go [hunk 1]
```diff
@@ -64,6 +68,8 @@ func (d *dexterOIDC) initialize() error {
 	kubeConfigDefaultPath := filepath.Join(usr.HomeDir, ".kube", "config")
 
 	// setup commandline flags
+	AuthCmd.PersistentFlags().StringVarP(&d.endpoint, "endpoint", "e", "azure", "OIDC-providers: google or azure")
```
[the remark this answers is not in the document]
Comment (cblims) [reply cmd/auth.go]: sorry, my mistake. Had it set to azure for easier testing.

**Comment by cblims** [comment]
Not sure what happens with travis. What is it complaining about?

File: cmd/auth.go [hunk 2]
```diff
@@ -88,22 +94,29 @@ func (d *dexterOIDC) createOauth2Config() error {
...
+	case "azure":
+		d.Oauth2Config.Endpoint = microsoft.AzureADEndpoint(oidcData.azureTenant)
+		d.Oauth2Config.Scopes = []string{oidc.ScopeOpenID, oidc.ScopeOfflineAccess, "email"}
+	case "google":
+		d.Oauth2Config.Endpoint = google.Endpoint
+		d.Oauth2Config.Scopes = []string{oidc.ScopeOpenID, "profile", "email"}
+	default:
+		return errors.New(fmt.Sprintf("unsupported endpoint: %s", oidcData.endpoint))
+	}
+		d.Oauth2Config.Endpoint = microsoft.AzureADEndpoint(oidcData.azureTenant)
```
Comment (dkerwin) [inline cmd/auth.go]: Thanks for accepting my code suggestions. Unfortunately github is not super flexible when it comes to multiline/block changes. Line #114 - #119 in the diff would need to be removed as well. Sorry

Comment (cblims) [reply cmd/auth.go]: sorry, didn't see them until I pulled locally. Should be fixed now. What do I do to satisfy buildstep 3 in Travis?

**Comment by dkerwin** [comment]
I have updated the travis build config. I'm pretty sure that the build will succeed if you rebase your branch.

**Comment by cblims** [comment]
make vet still fails.

**Comment by dkerwin** [comment]
golang.org/x/oauth2/google and others are not part of the go mod. Please ensure that the go module config is uptodate. If not the build with 1.12 will fail

**Comment by cblims** [comment]
Long way, and steep learning curve. go.mod had the necessary module, but vendor folder didn't have hit. I thought that go1.2 didn't use this folder...

Ready for merge?

File: cmd/auth.go [hunk 3]
```diff
@@ -179,42 +186,33 @@ func (d *dexterOIDC) callbackHandler(w http.ResponseWriter, r *http.Request) {
...
+	defer d.k8sMutex.Unlock()
 
-	uri.RawQuery = q.Encode()
+	idToken := token.Extra("id_token").(string)
 
-	resp, err := d.httpClient.Get(uri.String())
+	parsed, err := jwt.ParseSigned(idToken)
 	if err != nil {
-		return "", err
+		panic(err)
```
[the remark this answers is not in the document]
Comment (cblims) [reply cmd/auth.go]: Agree, printing error message.

File: cmd/auth.go [hunk 4]
```diff
@@ -64,6 +68,8 @@ func (d *dexterOIDC) initialize() error {
 	kubeConfigDefaultPath := filepath.Join(usr.HomeDir, ".kube", "config")
 
 	// setup commandline flags
+	AuthCmd.PersistentFlags().StringVarP(&d.endpoint, "endpoint", "e", "google", "OIDC-providers: google or azure")
+	AuthCmd.PersistentFlags().StringVarP(&d.azureTenant, "tenant", "t", "common", "Your azure tenant (default: common)")
```
[the remark this answers is not in the document]
Comment (cblims) [reply cmd/auth.go]: fixed

**Comment by dkerwin** [comment]
Thank you for all the time and effort you put into this! Much appreciated!

**Pull request merged**